core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -289,19 +311,24 @@ int readVersionHint() {
       return Integer.parseInt(in.readLine().replace("\n", ""));
 
     } catch (Exception e) {
-      LOG.warn("Error reading version hint file {}", versionHintFile, e);
+      LOG.debug("Error reading version hint file {}", versionHintFile, e);
       try {
-        if (getMetadataFile(1) != null) {
-          // We just assume corrupted metadata and start to read from the first version file
-          return 1;
+        // List the metadata directory to find the version files, and try to recover the max available version
+        FileStatus[] files = fs.listStatus(metadataRoot(), name -> VERSION_PATTERN.matcher(name.getName()).matches());
```

Comment:
This will list all the files in the metadata directory, right? It's just that we're filtering on the client for the files matching the version naming pattern.

Reply:
Sadly, yes. I did not find a better solution where the filter is pushed to server side.
Any ideas how to do it more efficiently?